lib/icingadb/icingadb-objects.cpp

```diff
@@ -2282,7 +2292,92 @@ void IcingaDB::SendAcknowledgementCleared(const Checkable::Ptr& checkable, const
 		xAdd.emplace_back(removedBy);
 	}
 
-	m_Rcon->FireAndForgetQuery(std::move(xAdd), Prio::History);
+	m_HistoryBulker.ProduceOne(std::move(xAdd));
+}
+
+void IcingaDB::ForwardHistoryEntries()
+{
+	using clock = std::chrono::steady_clock;
+
+	const std::chrono::seconds logInterval (10);
+	auto nextLog (clock::now() + logInterval);
+
+	auto logPeriodically ([this, logInterval, &nextLog]() {
+		if (clock::now() > nextLog) {
+			nextLog += logInterval;
+
+			auto size (m_HistoryBulker.Size());
+
+			Log(size > m_HistoryBulker.GetBulkSize() ? LogInformation : LogNotice, "IcingaDB")
+				<< "Pending history queries: " << size;
+		}
+	});
+
+	for (;;) {
+		logPeriodically();
+
+		auto haystack (m_HistoryBulker.ConsumeMany());
+
+		if (haystack.empty()) {
+			if (!GetActive()) {
+				break;
+			}
+
+			continue;
+		}
+
+		uintmax_t attempts = 0;
+
+		auto logFailure ([&haystack, &attempts](const char* err = nullptr) {
+			Log msg (LogNotice, "IcingaDB");
+
+			msg << "history: " << haystack.size() << " queries failed temporarily (attempt #" << ++attempts << ")";
+
+			if (err) {
+				msg << ": " << err;
+			}
+		});
+
+		for (;;) {
+			logPeriodically();
+
+			if (m_Rcon && m_Rcon->IsConnected()) {
+				auto future (m_Rcon->AsyncGetResultsOfQueries(haystack, Prio::History));
+				bool timedOut = false;
+
+				while (future.wait_for(std::chrono::minutes(1)) == std::future_status::timeout) {
+					if (!GetActive()) {
+						logFailure("timeout");
+						timedOut = true;
+						break;
+					}
+				}
```

Comment:
> Theoretically this could delay the shutdown by more than a minute in total

OK, but we **want** to flush all already buffered if Redis is responding, right?

Reply:
In general yes, but not at any cost. And the question is where do we stop. If we wanted to write history at all cost, we couldn't just stop if the Redis connections fails once either. I think it's a good idea to have a fixed limit there, if 1 minute is the perfect limit is up to discussion, so far it just sounded like a good value that not too short but also not too long either.